File: data/spark/src/daily_data/stock_price/calculate_rs.py
import psycopg2
from datetime import datetime, timedelta
import yfinance as yf

#check if market is opened
def check_market_open():
    end = datetime.now()
    start = end - timedelta(1)
    is_market_open = yf.download("SPY", start=start, end=end, group_by='ticker', auto_adjust=True)
    print(is_market_open)

    if is_market_open.empty:
        print("market is closed ytd")
        exit()

def calculate_rs(conn):
    # create a cursor
    cur = conn.cursor()

    cur.execute("""REFRESH MATERIALIZED VIEW stock_historic_prices;""")

    cur.execute("select * from stock_historic_prices where ticker='IWV';")

    russel = cur.fetchall()[0]

    russel_perf = performance(russel)

    # print("russel performance: ", russel_perf)

    cur.execute("select * from stock_historic_prices;")

    stock_list = cur.fetchall()

    tuples = []

    for stock in stock_list:
        stock_perf = performance(stock)
        date_id = stock[1]
        stock_id = stock[0]
        off_year_high = -float(percentChg(stock[4], stock[3]))
        relative_strength = float(percentChg(stock_perf, russel_perf))
        tuple = (date_id, stock_id, off_year_high, relative_strength)
        tuples.append(tuple)

    try:
        query_str = to_query_str(tuples)
        cur.execute("delete from temp_stock_rs;")
        cur.execute("INSERT INTO temp_stock_rs (date_id, stock_id, off_year_high, relative_strength) VALUES " + query_str) 

        print('Inserted into temp_stock_rs')

        conn.commit()

    except (Exception, psycopg2.DatabaseError) as error:
        print(f"Error: {error}")
        conn.rollback()
        cur.close()
        return 

    try:
        cur.execute("""INSERT INTO stock_rs (date_id, stock_id, rs_rating)
                    (select date_id, stock_id,
			        round(percent_rank() OVER (ORDER BY relative_strength) * 100) as rs_rating
			        from temp_stock_rs order by stock_id) on conflict(stock_id, date_id) 
	 			    DO UPDATE set rs_rating = EXCLUDED.rs_rating, updated_at = NOW()""")
                     
        print('Inserted into stock_rs')

        conn.commit()

    except (Exception, psycopg2.DatabaseError) as error:
        print(f"Error: {error}")
        conn.rollback()
        cur.close()
        return 

    cur.close()


def industry_rs(conn):
    cur = conn.cursor()

    cur.execute("""REFRESH MATERIALIZED VIEW industry_rs_view;""")

    cur.execute("""INSERT INTO industry_rs (date_id, industry_id, rs_rating, ranking)
                (select date_id, id as "industry_id",
		        round(industry_rs) as "rs_rating",
		        rank () over (order by industry_rs desc) as "ranking"
		        from industry_rs_view order by id) on conflict(industry_id, date_id) 
                DO UPDATE set rs_rating = EXCLUDED.rs_rating, ranking = EXCLUDED.ranking, updated_at = NOW()""")

    print('Inserted into industry_rs')

    cur.execute("""REFRESH MATERIALIZED VIEW screeners;""")

    conn.commit()

    cur.close()

def percentChg(now, before):
    return ((now - before) / abs(before)) * 100

def performance(stock):
    perf = ['', '', '', '', '']
    for i in range(5):
        if stock[i + 5]:
            perf[i] = float(percentChg(stock[4], stock[i + 5]))
        else: 
            perf[i] = perf[i - 1]
    return 0.15 * perf[0] + 0.25 * perf[1] + 0.2 * perf[2] + 0.2 * perf[3] + 0.2 * perf[4]


def to_query_str(tuples):
    args_str = ""
    for tuple in tuples:
        args_str += str(tuple)
        args_str += ","
    return args_str[:-1]

def close_connection(conn):
    if conn is not None:
        conn.close()
        print('Database connection closed.')

def main(config, section: str):

    check_market_open()

    print("connecting to", section)
    
    # read connection parameters
    params = config(section)

    # connect to the PostgreSQL server
    conn = psycopg2.connect(**params)

    calculate_rs(conn)

    industry_rs(conn)

    close_connection(conn)

if __name__ == '__main__':
    import os
    import sys
    os.chdir("/opt/bitnami/spark/src/")
    sys.path.append('psql_config/')
    from config import config

    main(config, 'postgresql-data')
    main(config, 'postgresql-web')
    exit()

File: backend/util/calculate_rs.ts
import Knex from "knex";
import knexConfigs from "../knexfile";
import env from "./env";

const knex = Knex(knexConfigs[env.NODE_ENV]);

async function init() {
	await knex.raw(`REFRESH MATERIALIZED VIEW stock_historic_prices`);
	await calculateRs();
	await knex.raw(`REFRESH MATERIALIZED VIEW industry_rs_view`);
	await industryRs();
	await knex.raw(`REFRESH MATERIALIZED VIEW screeners`);
	knex.destroy();
}

init();

async function calculateRs() {
	let stockArr: StockPrice[] = await knex.select("*").from("stock_historic_prices");

	let russel: StockPrice = (await knex.select("*").from("stock_historic_prices").where("ticker", "IWV"))[0];

	const russelPerf = performance(russel);

	console.log(russelPerf);

	let stockRsArr = [];

	for (let stock of stockArr) {
		let obj: StockRS = {
			date_id: 0,
			stock_id: 0,
			off_year_high: 0,
			relative_strength: 0,
		};

		let stockPerf = performance(stock);
		obj.date_id = stock.date_id;
		obj.stock_id = stock.stock_id;
		obj.relative_strength = Number(percentChg(stockPerf, russelPerf).toFixed(2));
		obj.off_year_high = -Number(percentChg(stock.price1, stock.year_high).toFixed(2));
		stockRsArr.push(obj);
	}

	// console.log(stockRsArr);

	await knex("temp_stock_rs").del()

	await knex.insert(stockRsArr).into("temp_stock_rs");

	let resultArr = (
		await knex.raw(
			/*sql*/
			`select stock_id, date_id,
			round(percent_rank() OVER (
			ORDER BY relative_strength) * 100) as rs_rating
			from temp_stock_rs
			order by stock_id`
		)
	).rows;

	console.log(resultArr[0]);

	await knex.insert(resultArr).into("stock_rs");
}

async function industryRs() {
	console.log("start calculating industry rs");

	let industryRsArr: inductryRS[] = (
		await knex.raw(
			/*sql*/
		`select date_id, id as "industry_id",
		round(industry_rs) as "rs_rating",
		rank () over (
		order by industry_rs desc) as "ranking"
		from industry_rs_view
		order by id`
		)
	).rows;

	await knex.insert(industryRsArr).into("industry_rs");
	console.log("finish calculating industry rs");
}

function percentChg(now: number, before: number) {
	return ((now - before) / Math.abs(before)) * 100;
}

function performance(stock: StockPrice) {
	let perf: any[] = [];
	for (let i = 0; i < 5; i++) {
		if (stock[`price${i + 2}`]) {
			perf[i] = percentChg(stock.price1, stock[`price${i + 2}`]);
		} else {
			perf[i] = perf[i - 1];
		}
	}
	return 0.15 * perf[0] + 0.25 * perf[1] + 0.2 * perf[2] + 0.2 * perf[3] + 0.2 * perf[4];
}

type StockPrice = {
	date_id: number;
	stock_id: number;
	ticker: string;
	year_high: number;
	price1: number;
	price2: number;
	price3: number;
	price4: number;
	price5: number;
	price6: number;
};

type StockRS = {
	date_id: number;
	stock_id: number;
	off_year_high: number;
	relative_strength?: number;
	rs_rating?: number;
};

type inductryRS = {
	date_id: number;
	industry_id: number;
	rs_rating: number;
	ranking: number;
}